recipe-app/src/RecipeForm.js

```diff
@@ -0,0 +1,123 @@
+import React from "react";
+import styled from "styled-components";
+
+import useInputState from "./hooks/useInputState";
+import useStringListInputState from "./hooks/useStringListInputState";
+import NewIngredientForm from "./NewIngredientForm";
+import {createRecipeData} from "./utils/recipeAPIUtils";
+
+const StyledRecipeForm = styled.form`
+    border: black 5px double;
+    padding: 5px;
+    margin: 5px;
+    background: lightgray;
+    width: 500px;
+    text-align: left;
+`
+
+const StyledDiv = styled.div`
+    display: flex;
+`
+
+const StyledParagraph = styled.p`
+    display: flex;
+`
+
+const StyledIngredientList = styled.div`
+    display: inline;
+    align-items: flex-start;
+    flex-wrap: wrap;
+    margin: 0px;
+    gap: 0px;
+`
+
+const StyledLabel = styled.label`
+    width: 100px;
+`
+
+const StyledInput = styled.input`
+    width: 90%;
+    margin: 2px;
+`
+
+function RecipeForm(props) {
+    let recipe = props.recipe;
+
+    if (!recipe) {
+        recipe = {
+            "name": "",
+            "description": "",
+            "ingredients": []
+        };
+    }
+
+    const [name, handleNameChange] = useInputState(recipe.name);
+    const [description, handleDescriptionChange] =
+        useInputState(recipe.description);
+    const [ingredients, addBlankIngredient, handleIngredientChange, removeIngredient] = useStringListInputState(recipe.ingredients);
+
+    const buildIngredientsForms = () => {
+        return ingredients.map((name, idx) => <NewIngredientForm
+            name={name} idx={idx} key={idx}
+            handleNameChange={handleIngredientChange}
+            removeIngredient={removeIngredient}/>);
+    }
+
+    const handleCreate = (e) => {
+        e.preventDefault();
+
+        createRecipeData({
+            name: name,
+            description: description,
+            ingredients: ingredients
+        }, true, props);
```

Comment:
I see that you're passing the props so to do `history.push`. API methods should just call the API, the redirect should live within the component: it improves testability and the concerns are better separated.
```tsx
const handleCreate = async (e) => {
  /* etc */
  const recipe = await createRecipeData()
  history.push(`/recipe/${recipe.id}`)
```